File: cloud-collect-core/src/main/java/com/windea/demo/cloudcollect/core/controller/CommentController.java
package com.windea.demo.cloudcollect.core.controller;

import com.windea.demo.cloudcollect.core.domain.entity.Comment;
import com.windea.demo.cloudcollect.core.domain.model.JwtUserDetails;
import com.windea.demo.cloudcollect.core.service.CommentService;
import io.swagger.annotations.*;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.validation.BindingResult;
import org.springframework.web.bind.annotation.*;

import javax.validation.Valid;
import java.security.Principal;

/**
 * 评论的控制器。
 */
@Api("评论")
@RestController
@RequestMapping("/comment")
@CrossOrigin
public class CommentController {
	private final CommentService service;

	public CommentController(CommentService service) {
		this.service = service;
	}


	@ApiOperation("创建自己的评论。")
	@ApiImplicitParams({
		@ApiImplicitParam(name = "collectId", value = "评论的收藏的id", required = true),
		@ApiImplicitParam(name = "comment", value = "新的评论", required = true)
	})
	@PostMapping("/create")
	public void create(@RequestParam Long collectId,
		@RequestBody @Valid Comment comment, BindingResult bindingResult, Principal principal) {
		var user = ((JwtUserDetails) principal).getDelegateUser();
		service.create(collectId, comment, user);
	}

	@ApiOperation("创建自己的评论，回复某一评论。")
	@ApiImplicitParams({
		@ApiImplicitParam(name = "collectId", value = "评论的收藏的id", required = true),
		@ApiImplicitParam(name = "replyToCommentId", value = "回复的评论的id", required = true),
		@ApiImplicitParam(name = "comment", value = "新的评论", required = true)
	})
	@PostMapping("/reply")
	public void reply(@RequestParam Long collectId, @RequestParam Long replyToCommentId,
		@RequestBody @Valid Comment comment, BindingResult bindingResult, Principal principal) {
		var user = ((JwtUserDetails) principal).getDelegateUser();
		service.reply(collectId, replyToCommentId, comment, user);
	}

	@ApiOperation("删除自己的评论。")
	@ApiImplicitParams({
		@ApiImplicitParam(name = "id", value = "id", required = true, paramType = "path")
	})
	@DeleteMapping("/{id}")
	public void delete(@PathVariable Long id) {
		service.delete(id);
	}

	@ApiOperation("得到某一评论。")
	@ApiImplicitParams({
		@ApiImplicitParam(name = "id", value = "id", required = true, paramType = "path")
	})
	@GetMapping("/{id}")
	public Comment get(@PathVariable Long id) {
		return service.get(id);
	}

	@ApiOperation("分页得到回复某一评论的所有评论。")
	@ApiImplicitParams({
		@ApiImplicitParam(name = "id", value = "id", required = true, paramType = "path"),
		@ApiImplicitParam(name = "pageable", value = "分页和排序", required = true)
	})
	@GetMapping("/{id}/replyByCommentPage")
	public Page<Comment> getReplyByCommentPage(@PathVariable Long id, @RequestParam Pageable pageable) {
		return service.getReplyByCommentPage(id, pageable);
	}

	@ApiOperation("得到回复某一评论的评论数量。")
	@ApiImplicitParams({
		@ApiImplicitParam(name = "id", value = "id", required = true, paramType = "path")
	})
	@GetMapping("/{id}/replyByCommentCount")
	public Long getReplyByCommentCount(@PathVariable Long id) {
		return service.getReplyByCommentCount(id);
	}

	@ApiOperation("分页得到所有评论。")
	@ApiImplicitParams({
		@ApiImplicitParam(name = "pageable", value = "分页和排序", required = true)
	})
	@GetMapping("/findAll")
	public Page<Comment> findAll(@RequestParam Pageable pageable) {
		return service.findAll(pageable);
	}

	@ApiOperation("分页查询某一收藏的所有评论。")
	@ApiImplicitParams({
		@ApiImplicitParam(name = "collectId", value = "收藏的id", required = true),
		@ApiImplicitParam(name = "pageable", value = "分页和排序", required = true)
	})
	@GetMapping("/findByCollect")
	public Page<Comment> findByCollect(@RequestParam Long collectId, @RequestParam Pageable pageable) {
		return service.findByCollect(collectId, pageable);
	}
}

File: cloud-collect-core/src/main/kotlin/com/windea/demo/cloudcollect/core/controller/CommentController.kt
@file:Suppress("UNUSED_PARAMETER")

package com.windea.demo.cloudcollect.core.controller

import com.windea.demo.cloudcollect.core.domain.entity.*
import com.windea.demo.cloudcollect.core.service.*
import com.windea.demo.cloudcollect.core.validation.group.*
import io.swagger.annotations.*
import org.springframework.data.domain.*
import org.springframework.security.access.prepost.*
import org.springframework.validation.*
import org.springframework.validation.annotation.*
import org.springframework.web.bind.annotation.*

@Api("评论")
@RestController
@RequestMapping("/comment")
@CrossOrigin
class CommentController(
	private val commentService: CommentService
) {
	@ApiOperation("创建自己的评论。")
	@PostMapping("/create")
	@PreAuthorize("isAuthenticated()")
	fun create(@RequestBody @Validated(Create::class) comment: Comment, bindingResult: BindingResult) {
		commentService.create(comment)
	}
	
	@ApiOperation("回复某一评论。")
	@PostMapping("/reply")
	@PreAuthorize("isAuthenticated()")
	fun reply(@RequestBody @Validated(Create::class) comment: Comment, bindingResult: BindingResult) {
		commentService.reply(comment)
	}
	
	@ApiOperation("删除自己的评论。")
	@DeleteMapping("/{id}")
	@PreAuthorize("isAuthenticated()")
	fun deleteById(@PathVariable id: Long) {
		commentService.deleteById(id)
	}
	
	@ApiOperation("根据id得到某一评论。")
	@GetMapping("/{id}")
	fun findById(@PathVariable id: Long): Comment {
		return commentService.findById(id)
	}
	
	@ApiOperation("得到所有评论。")
	@GetMapping("/findAll")
	fun findAll(pageable: Pageable): Page<Comment> {
		return commentService.findAll(pageable)
	}
	
	@ApiOperation("根据收藏id查询所有评论。")
	@GetMapping("/findAllByCollectId")
	fun findAllByCollectId(@RequestParam collectId: Long, pageable: Pageable): Page<Comment> {
		return commentService.findAllByCollectId(collectId, pageable)
	}
	
	@ApiOperation("根据发起用户id查询所有评论。")
	@GetMapping("/findAllBySponsorByUserId")
	fun findAllBySponsorByUserId(@RequestParam sponsorByUserId: Long, pageable: Pageable): Page<Comment> {
		return commentService.findAllBySponsorByUserId(sponsorByUserId, pageable)
	}
	
	@ApiOperation("根据回复评论id查询所有评论。")
	@GetMapping("/findAllByReplyToCommentId")
	fun findAllByReplyToCommentId(@RequestParam replyToCommentId: Long, pageable: Pageable): Page<Comment> {
		return commentService.findAllByReplyToCommentId(replyToCommentId, pageable)
	}
}
